feat: add `pyroscope.receive_http` component for profile handling (#1886)
grafana/alloy: docs/sources/reference/components/pyroscope/pyroscope.receive_http.md

---
canonical: https://grafana.com/docs/alloy/latest/reference/components/pyroscope/pyroscope.receive_http/
aliases:
  - ../pyroscope.receive_http/ # /docs/alloy/latest/reference/components/pyroscope.receive_http/
description: Learn about pyroscope.receive_http
title: pyroscope.receive_http
---

# pyroscope.receive_http

`pyroscope.receive_http` listens for HTTP requests containing profiles and forwards them to other components capable of receiving profiles.

> The HTTP API exposed is compatible with Pyroscope's [HTTP ingest API](https://grafana.com/docs/pyroscope/latest/configure-server/about-server-api/). This allows `pyroscope.receive_http to act as a proxy for Pyroscope profiles, enabling flexible routing and distribution of profile data.

Reviewer: ```suggestion
The HTTP API exposed is compatible with the Pyroscope [HTTP ingest API](https://grafana.com/docs/pyroscope/latest/configure-server/about-server-api/).
This allows `pyroscope.receive_http` to act as a proxy for Pyroscope profiles, enabling flexible routing and distribution of profile data.
```

Name              | Type          | Description                                     | Default | Required
------------------|---------------|-------------------------------------------------|---------|---------
`forward_to` | `list(ProfilesReceiver)` | List of receivers to send profiles to. |         | yes

## Blocks

The following blocks are supported inside the definition of `pyroscope.receive_http`:

Hierarchy | Name | Description                                        | Required
----------|------|----------------------------------------------------|---------
`http`    | `http` | Configures the HTTP server that receives requests. | no

### http

The `http` block configures the HTTP server.

You can use the following arguments to configure the `http` block. Any omitted fields take their default values.

Name                   | Type       | Description                                                                                                      | Default  | Required
-----------------------|------------|------------------------------------------------------------------------------------------------------------------|----------|---------

> `conn_limit`           | `int`      | Maximum number of simultaneous HTTP connections. Defaults to no limit.                                           | `0`      | no

Reviewer 2: We have been burnt on no limit in the past, thoughts on making this a a high reasonable limit for default?

Author: I can add a default limit. Do you have any data/knowledge that could help us set a reasonable maximum limit?

Reviewer 2: 100? We want to prevent thousands unless specified.

> `listen_address`       | `string`   | Network address on which the server listens for new connections. Defaults to accepting all incoming connections. | `""`     | no

Reviewer 2: Is this only HTTP or is HTTPS also supported? From the verbiage and underlying server I assume only HTTP?

Author: From Alloy's perspective, I believe this is acceptable since we have other receiver components (like Loki's [API component](https://github.com/grafana/alloy/blob/main/internal/component/loki/source/api/api.go#L32)) using HTTP-only. From our perspective, Pyroscope SDKs don't require TLS for pushing profiles, so we're fine on this side. I'd say we can add HTTPS support in the future if needed. 
Does this make sense to you?

`server_write_timeout` | `duration` | Write timeout for HTTP server.                                                                                   | `"30s"`  | no

## Exported fields

`pyroscope.receive_http` does not export any fields.

## Component health

`pyroscope.receive_http` is reported as unhealthy if it is given an invalid configuration.

## Example
This example creates a `pyroscope.receive_http` component which starts an HTTP server listening on `0.0.0.0` and port `9999`. The server receives profiles and forwards them to multiple `pyroscope.write` components, which write these profiles to different HTTP endpoints.
```alloy
// Receives profiles over HTTP
pyroscope.receive_http "default" {
  http {
    listen_address = "0.0.0.0"
    listen_port = 9999
  }
  forward_to = [pyroscope.write.staging.receiver, pyroscope.write.production.receiver]
}

// Send profiles to a staging Pyroscope instance
pyroscope.write "staging" {
  endpoint {
    url = "http://pyroscope-staging:4040"
  }
}

// Send profiles to a production Pyroscope instance
pyroscope.write "production" {
  endpoint {
    url = "http://pyroscope-production:4040"
  }
}
```

> Note: This example demonstrates forwarding to multiple `pyroscope.write` components. Be aware that this configuration will duplicate the received profiles, sending a copy to each configured `pyroscope.write` component.

Reviewer: ```suggestion
{{< admonition type="note" >}}
This example demonstrates forwarding to multiple `pyroscope.write` components.
This configuration will duplicate the received profiles and send a copy to each configured `pyroscope.write` component.
{{< /admonition >}}
```
